### PR TITLE
Integrate C++ kernels for 4-bit & 2-bit MatMul

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do ? 
This PR implements C++ udqmm kernel for 4-bit and 2-bit MatMul. In this kernel, we do the following steps: unpack + dequantize + mm where mm is the only parallelized operation for now.  We also implement the c++ version of `ungroup`

cc @SunMarc @dacorvo

### CODE REVIEW

File: quanto/library/ext/cpp/udqmm.cpp [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+#include "unpack.h"
+
+#include <iostream>
+#include <torch/extension.h>
+
+using namespace std;
+
+torch::Tensor udqmm(torch::Tensor &input, torch::Tensor &weights, torch::Tensor& scale, torch::Tensor& zeropoint, int axis, int bits, torch::IntArrayRef orig_shape) {
+    torch::Tensor unpacked_weights = unpack(weights, bits);
+    torch::Tensor dq_output = (unpacked_weights.to(torch::kInt8) - zeropoint.to(torch::kInt8)) * scale;
```
Comment (dacorvo) [inline quanto/library/ext/cpp/udqmm.cpp:11]: Ok to cast the unpacked_weigths since they are uint8 (we could also change the unpack method to do the cast there, making it part of the contract). We should only assert and err on zeropoint though, using TORCH_CHECK.

File: quanto/library/ext/cpp/udqmm.cpp [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+    torch::Tensor ungrouped_output;
+
+    // Ungroup TODO : put on its own function
```
Comment (dacorvo) [inline quanto/library/ext/cpp/udqmm.cpp:15]: Definitely. And this should be tested independently like unpack.

File: quanto/library/python/udqmm.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    weights: torch.Tensor,
+    scale: torch.Tensor,
+    zeropoint: torch.Tensor,
+    axis: int,
+    bits: int,
+    orig_shape: torch.Size,
+):
+    unpacked_weights = torch.ops.quanto.unpack(weights, bits)
+    shifted_weights = unpacked_weights.to(torch.int8) - zeropoint.to(torch.int8)
+    scaled_weights = shifted_weights * scale
```
Comment (dacorvo) [inline quanto/library/python/udqmm.py:18]: FYI, this will actually promote shifted_weigths to float16 then do the multiplication.
Since I had issues with these kind of implicit operation, maybe we could do that explicitly:
```
scaled_weights = shifted_weights.to(scale.dtype) * scale
```

File: test/library/test_mm.py [hunk 4]
```diff
@@ -1,7 +1,13 @@
+from contextlib import nullcontext
+
 import pytest
 import torch
 from helpers import random_tensor
 
+from quanto.library import disable_extensions
+from quanto.tensor.core import group, ungroup
+from quanto.tensor.packed import pack_weights, unpack_weights
```
Comment (dacorvo) [inline test/library/test_mm.py:9]: Agree we should not have two unpack methods. Here you could use torch.ops.quanto.unpack instead.

**Review by dacorvo — COMMENTED** [review]
Looks really really good ... I left a few comments. 
I am wondering how we can pass the qlinear tests though since I thought the packed weights would be transposed when passed to `mm` and their axis would be 1 which is not supported yet ...

File: quanto/library/ext/cpp/udqmm.cpp (hunk 1 above)
Comment (SunMarc) [reply quanto/library/ext/cpp/udqmm.cpp:11]: It will probably be cleaner do the cast in the unpack. So that we have the same dtype for `unpacked_weights` and `zeropoint` ! LMK what you think. If yes, should we do that in another PR ?

File: quanto/library/python/udqmm.py (hunk 3 above)
Comment (SunMarc) [reply quanto/library/python/udqmm.py:18]: Good to know !

File: quanto/library/ext/cpp/udqmm.cpp (hunk 1 above)
Comment (dacorvo) [reply quanto/library/ext/cpp/udqmm.cpp]: I thought about that a little bit and I am inclined to keep returning an unsigned int.
The unpacked weights really are unsigned integer in the range [0, 2**(bits -1)], and the
only reason for the cast is the subtraction of a signed int.
Now regarding the cast, I think we could simply drop it, as I realized pytorch is doing an implicit type promotion to `int16` because of the subtraction:
```python
>>> a = torch.tensor(255, dtype=torch.uint8).to('mps')
>>> b = torch.tensor(1, dtype=torch.int8).to('mps')
>>> a - b
tensor(254, device='mps:0', dtype=torch.int16)
```

Comment (dacorvo) [reply quanto/library/ext/cpp/udqmm.cpp]: The reason why pytorch is doing that is probably because the first term might become negative, and change the result:
```python
>>> a.to(torch.int8) - b
tensor(-2, device='mps:0', dtype=torch.int8)
```
In our case we know the first term will stay positive during the cast, so it is safe, and avoids the materialization of a large `int16` intermediate tensor.

File: bench/library/benchmark.py [hunk 5]
```diff
@@ -109,12 +121,12 @@ def elapsed_time(self, other):
...
-        torch.float32, torch.int8, False, device
-    ),
-    "unpack_2bit": lambda device: get_unpack_bench(2, device),
-    "unpack_4bit": lambda device: get_unpack_bench(4, device),
+    # "dqmm_w8a16": lambda device: get_dqmm_bench(torch.float16, device),
+    # "quantize_symmetric_fp32_int8_per_tensor": lambda device: get_quantize_symmetric_bench(
+    #     torch.float32, torch.int8, False, device
+    # ),
+    # "unpack_2bit": lambda device: get_unpack_bench(2, device),
+    # "unpack_4bit": lambda device: get_unpack_bench(4, device),
```
Comment (dacorvo) [inline bench/library/benchmark.py]: You can pass just the bench you want as parameter to the script ... just sayin ...

Comment (SunMarc) [reply bench/library/benchmark.py]: Haha, indeed !

File: test/library/test_mm.py (hunk 4 above)
Comment (SunMarc) [reply test/library/test_mm.py]: Done !

File: quanto/library/python/udqmm.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+) -> torch.Tensor:
+    # we transpose it back, so it is simpler to unpack since we have the pack + transposed weights
+    weights = weights.transpose(0, 1)
+    unpacked_weights = torch.ops.quanto.unpack(weights, bits)
+    # TODO : we should proably add that in unpack with with arg unpacked_shape.
+    # Depends if the weights have been transposed or not
+    # if not transposed, we need to do unpacked_weights[: unpacked_shape[0]]
+    unpacked_weights = unpacked_weights[: unpacked_shape[1]]
+    # transpose back
+    unpacked_weights = unpacked_weights.transpose(0, 1)
```
Comment (SunMarc) [inline quanto/library/python/udqmm.py:23]: When we call the linear layer, the Qbitstensor gets transposed and we end up with transposed weights in the udqmm function. This is quite troublesome for the `unpacking` step since don't have an variable to indicate (e.g. axis for the `ungroup` ) when we should transpose again our weights to unpack them correctly. Do you have any solution to fix this @dacorvo ? We can maybe add an attribute in `PackedTensor` that tell us if the data was transposed, so that we know how to unpack it. 

This is the reason why the test_mm.py are not passing yet. For the test_qlinear.py (where we have transposed weights), there is no issue because of the fix I did above.

Comment (dacorvo) [reply quanto/library/python/udqmm.py:23]: When the weights are transposed, the axis is also modified, so all the information should be available to the caller. I am not sure to see what the issue is here, because the flow has been validated in python already.

**Comment by dacorvo** [comment]
Before reviewing, it can you rebase your branch and squash the commits that are related to one another (like a fix to another commit).

Rebase on main:

The last commit from `main` you are using being 3bddd0a:
```
$ git rebase 3bddd0a --onto origin/main
```
You will need to solve conflicts (because at some point you merged an obsolete `main`, which makes things more difficult here).

Interactive rebase to cleanup redundant commits:

```
$ git rebase -i origin/main
```

This will open an editor with the following list of commits:

```
pick 7671993 v1 - does not support group wise?
pick 5ea42cb add grouped weights support for axis=0
pick c34443b refactor everything
pick 1c16b33 add python impl
pick 61b49e9 add benchmark script and fix python impl
pick f0bfa31 add dispatch to `QBitsTensor` and update benchmark
pick 2a8846e fix zero-point and orig_shape
pick cbd1922 add slicing
pick 0c485c7 uncomment
pick 81dc58a feat(cpp): check tensor type in udqmm with scalar_type
pick bc33442 feat(ungroup): add cpp and python implementation
pick 5bf6e18 feat(ungroup): add missing python impl
pick 02481a3 feat(PackedTensor): implement torch.ops.aten.t
pick 15734ad feat(QBitsTensor): implement torch.ops.aten.t
pick d3d280f feat(udqmm): fix issue with unpacking with transposed tensor
pick d1046df feat(udqmm): fix naming
```

Edit the command on the left to choose between `pick`, `squash` or `fixup` for each commit (you can also reorder them).
The difference between `squash` and `fixup` is that when squashing you keep the message.

An obvious example based on your commits:

```...
pick bc33442 feat(ungroup): add cpp and python implementation
fixup 5bf6e18 feat(ungroup): add missing python impl
...
```

File: quanto/library/python/udqmm.py (hunk 6 above)
Comment (SunMarc) [reply quanto/library/python/udqmm.py:23]: The issue is that the packing/unpacking is not dependent on the `axis`.

**Comment by SunMarc** [comment]
Hi @dacorvo, I've addressed the issue about the packing/unpacking and cleaned the commits !

**Comment by SunMarc** [comment]
I had the issue multiple time but it seems that the following test is flaky [test_quantize_symmetric[cpu-per-axis-float8-fp16-matrix]](https://github.com/huggingface/quanto/actions/runs/8268463403/job/22621393702).

**Comment by alejandroarmas** [comment]
Hey there @SunMarc @dacorvo I'm seeing that this PR was closed and wanted to understand how it relates to #107. I must have missed why the results were disappointing? Should someone in a position like mine scrap #107 and pivot to a different issue in the repo? Thank you in advance 😊 I just want to be as helpful as possible!
